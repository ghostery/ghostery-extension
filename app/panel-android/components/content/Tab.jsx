/**
 * Tab Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';

export default class Tab extends React.Component {
	handleTabClick = (event) => {
		event.preventDefault();
		this.props.onClick(this.props.tabIndex);
	}

	render() {
		return (
			<li className="tab-item">
				<a
					className={`tab-link ${this.props.linkClassName} ${this.props.isActive ? 'active' : ''}`}
					onClick={this.handleTabClick}
				>
					{this.props.tabLabel}
				</a>
			</li>
		);
	}
}

Tab.propTypes = {
	onClick: PropTypes.func,
	tabIndex: PropTypes.number,
	isActive: PropTypes.bool,
	tabLabel: PropTypes.string.isRequired,
	linkClassName: PropTypes.string.isRequired
};

Tab.defaultProps = {
	onClick: () => null,
	tabIndex: -1,
};

Tab.defaultProps = {
	isActive: false,
};
