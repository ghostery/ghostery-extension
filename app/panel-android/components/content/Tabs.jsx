/**
 * Tabs Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';

export default class Tabs extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			activeTabIndex: 0
		};
	}

	handleTabClick = (tabIndex) => {
		if (tabIndex === this.state.activeTabIndex) {
			return;
		}

		this.setState({
			activeTabIndex: tabIndex
		});
	}

	renderTabsNav = () => React.Children.map(this.props.children, (child, index) => React.cloneElement(child, {
		onClick: this.handleTabClick,
		tabIndex: index,
		isActive: index === this.state.activeTabIndex
	}));

	renderActiveTabContent = () => {
		const { children } = this.props;
		const { activeTabIndex } = this.state;
		if (children[activeTabIndex]) {
			return children[activeTabIndex].props.children;
		}
		return null;
	}

	render() {
		return (
			<div className="tabs-wrapper">
				<ul className="tabs-nav">
					{this.renderTabsNav()}
				</ul>
				<div className="tabs-active-content">
					{this.renderActiveTabContent()}
				</div>
			</div>
		);
	}
}
