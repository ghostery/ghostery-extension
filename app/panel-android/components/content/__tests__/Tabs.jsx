/**
 * Tabs Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Tabs from '../Tabs';
import Tab from '../Tab';

describe('app/panel-android/components/content/Tabs.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('Tabs component with 3 Tab children components', () => {
			const component = renderer.create(
				<Tabs>
					<Tab tabLabel="Tab 1 Label" linkClassName="tab-1-class">
						<div className="tab-1-content">Tab 1 Content</div>
					</Tab>

					<Tab tabLabel="Tab 2 Label" linkClassName="tab-2-class">
						<div className="tab-2-content">Tab 2 Content</div>
					</Tab>

					<Tab tabLabel="Tab 3 Label" linkClassName="tab-3-class">
						<div>Tab 3 Content Part I</div>
						<div>Tab 3 Content Part II</div>
						<div>Tab 3 Content Part III</div>
					</Tab>
				</Tabs>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Functionality tests mounted with Enzyme', () => {
		test('Tabs component with 3 Tab children components happy path', () => {
			const component = mount(
				<Tabs>
					<Tab tabLabel="Tab 1 Label" linkClassName="tab-1-class">
						<div className="tab-1-content">Tab 1 Content</div>
					</Tab>

					<Tab tabLabel="Tab 2 Label" linkClassName="tab-2-class">
						<div className="tab-2-content">Tab 2 Content</div>
					</Tab>

					<Tab tabLabel="Tab 3 Label" linkClassName="tab-3-class">
						<div>Tab 3 Content Part I</div>
						<div>Tab 3 Content Part II</div>
						<div>Tab 3 Content Part III</div>
					</Tab>
				</Tabs>
			);
			expect(component.find('.Tabs__component').length).toBe(1);
			expect(component.find('.Tabs__navigation').length).toBe(1);
			expect(component.find('.Tab__navigation_item').length).toBe(3);
			expect(component.find('.Tab__navigation_item.tab-1-class').length).toBe(0);
			expect(component.find('.Tab__navigation_item.Tab--active').length).toBe(1);
			expect(component.find('.Tab__navigation_link').length).toBe(3);
			expect(component.find('.Tab__navigation_link.Tab--active').length).toBe(1);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-1-class').length).toBe(1);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-2-class').length).toBe(0);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-3-class').length).toBe(0);
			expect(component.find('.Tabs__active_content').length).toBe(1);
			expect(component.find('.Tabs__active_content .tab-1-content').length).toBe(1);
			expect(component.find('.Tabs__active_content .tab-2-content').length).toBe(0);

			component.setState({ activeTabIndex: 1 });
			expect(component.find('.Tab__navigation_link.Tab--active.tab-1-class').length).toBe(0);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-2-class').length).toBe(1);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-3-class').length).toBe(0);
			expect(component.find('.Tabs__active_content .tab-1-content').length).toBe(0);
			expect(component.find('.Tabs__active_content .tab-2-content').length).toBe(1);

			component.setState({ activeTabIndex: 2 });
			expect(component.find('.Tab__navigation_link.Tab--active.tab-1-class').length).toBe(0);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-2-class').length).toBe(0);
			expect(component.find('.Tab__navigation_link.Tab--active.tab-3-class').length).toBe(1);
			expect(component.find('.Tabs__active_content .tab-1-content').length).toBe(0);
			expect(component.find('.Tabs__active_content .tab-2-content').length).toBe(0);
		});
	});
});
