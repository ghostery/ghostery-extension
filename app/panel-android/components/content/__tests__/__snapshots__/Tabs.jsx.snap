// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel-android/components/content/Tabs.jsx Snapshot tests with react-test-renderer Tabs component with 3 Tab children components 1`] = `
<div
  className="Tabs__component"
>
  <ul
    className="Tabs__navigation flex-container"
  >
    <li
      className="Tab__navigation_item flex-container align-center-middle Tab--active"
      onClick={[Function]}
    >
      <a
        className="Tab__navigation_link tab-1-class Tab--active"
      >
        Tab 1 Label
      </a>
    </li>
    <li
      className="Tab__navigation_item flex-container align-center-middle"
      onClick={[Function]}
    >
      <a
        className="Tab__navigation_link tab-2-class"
      >
        Tab 2 Label
      </a>
    </li>
    <li
      className="Tab__navigation_item flex-container align-center-middle"
      onClick={[Function]}
    >
      <a
        className="Tab__navigation_link tab-3-class"
      >
        Tab 3 Label
      </a>
    </li>
  </ul>
  <div
    className="Tabs__active_content"
  >
    <div
      className="tab-1-content"
    >
      Tab 1 Content
    </div>
  </div>
</div>
`;
